components/use-paid-mutation.js

```diff
@@ -23,39 +25,126 @@ export function usePaidMutation (mutation,
+    let response = originalResponse
+    let invoiceUsed = false
+
+    // ensures every invoice is used only once
+    const refreshInvoice = async () => {
+      if (invoiceUsed) {
+        const retry = await retryPaidAction({ variables: { invoiceId: parseInt(invoice.id) } })
+        response = retry.data?.retryPaidAction
+        invoice = response?.invoice
+        invoiceUsed = true
+      } else invoiceUsed = true
+    }
+
+    const cancelInvoice = async () => {
+      try {
+        invoiceUsed = true
+        await invoiceHelper.cancel(invoice)
+        console.log('old invoice canceled')
+      } catch (err) {
+        console.error('could not cancel old invoice', err)
+      }
+    }
+
+    // if anon we go straight to qr code
+    if (!me) {
+      await refreshInvoice()
+      if (!invoice) {
+        setInnerResult(r => addPayError(new Error('You must be logged in'), r))
+        throw new Error('You must be logged in')
+      }
+      await waitForQrPayment(invoice, null, { persistOnNavigate, waitFor })
+      return { invoice, response }
+    }
+
+    // we try every sender<->receiver combination
+    while (true) {
+      await refreshInvoice()
+      if (!invoice) return { invoice, response }
+
+      for (const wallet of usableWallets) {
+        try {
+          await waitForWalletPayment(invoice, waitFor, wallet)
+          console.log('paid with wallet', wallet.def.name)
+          return { invoice, response }
+        } catch (err) {
+          walletErrors.push(err)
+          console.log('could not pay with wallet', wallet.def.name, 'will try another one ...')
+        }
+      }
 
-  const waitForPayment = useCallback(async (invoice, { alwaysShowQROnFailure = false, persistOnNavigate = false, waitFor }) => {
-    let walletError
-    const start = Date.now()
+      // every wallet failed, but we can still retry, so let's ask for a better invoice
+      if (response.retriableWithWallet) {
+        console.log('could not pay with any wallet, will retry with a new invoice...')
+      } else {
+        console.log('could not pay with any wallet, and the invoice is not retriable, will fallback to another method')
+        break
+      }
+    }
+
+    // we try an internal payment
     try {
-      return await waitForWalletPayment(invoice, waitFor)
-    } catch (err) {
-      if (
-        (!alwaysShowQROnFailure && Date.now() - start > 1000) ||
-        err instanceof InvoiceCanceledError ||
-        err instanceof InvoiceExpiredError) {
-        // bail since qr code payment will also fail
-        // also bail if the payment took more than 1 second
-        // and cancel the invoice if it's not already canceled so it can be retried
-        invoiceHelper.cancel(invoice).catch(console.error)
-        throw err
+      console.log('could not pay with any wallet, will try with an internal payment...')
+      const retry = await retryPaidAction({ variables: { invoiceId: parseInt(invoice.id), prioritizeInternal: true } })
```

Comment:
In the previous step we failed to pay the action with whatever preferred method it had, so now we try to reprioritize FEE_CREDITS and REWARD_SATS so that they are tried first (if the action supports them).
If this fails to, we will get a new invoice out of this call, this invoice can be used in the qr code